File: frontend/shared/types/habits.ts
export const UNKNOWN = -1;
export const NO = 0;
export const YES_AUTO = 1;
export const YES_MANUAL = 2;
export const SKIP = 3;

export interface Habit {
  id: number;
  uuid: string;
  name: string;
  question: string;
  description: string;
  type: 0 | 1; // 0=yes/no, 1=numerical
  color: number;
  position: number;
  freq_num: number;
  freq_den: number;
  reminder_hour: number | null;
  reminder_min: number | null;
  reminder_days: number;
  target_type: 0 | 1; // 0=at least, 1=at most
  target_value: number;
  unit: string;
  created_at: string;
}

export interface HabitOverviewItem {
  habit: Habit;
  score: number;
  streak: number;
  entries: Record<string, number>;
  notes: Record<string, string>;
}

export interface EntryChange {
  value: number;
  score: number;
  streak: number;
  entries: Record<string, number>;
}

export interface HabitForm {
  name: string;
  question: string;
  description: string;
  type: 0 | 1;
  color: number;
  freq_num: number;
  freq_den: number;
  reminder_hour: number | null;
  reminder_min: number | null;
  reminder_days: number;
  target_type: 0 | 1;
  target_value: number;
  unit: string;
}

File: backend/app/api/schemas/habits.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class HabitBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    question: str = Field(default="", max_length=500)
    description: str = Field(default="", max_length=2000)
    type: int = Field(default=0, ge=0, le=1)
    color: int = Field(default=8, ge=0, le=19)
    freq_num: int = Field(default=1, ge=1, le=366)
    freq_den: int = Field(default=1, ge=1, le=366)
    reminder_hour: int | None = Field(default=None, ge=0, le=23)
    reminder_min: int | None = Field(default=None, ge=0, le=59)
    reminder_days: int = Field(default=127, ge=0, le=127)
    target_type: int = Field(default=0, ge=0, le=1)
    target_value: float = Field(default=0.0, ge=0)
    unit: str = Field(default="", max_length=50)


class HabitCreate(HabitBase):
    pass


class HabitPatch(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    question: str | None = Field(default=None, max_length=500)
    description: str | None = Field(default=None, max_length=2000)
    color: int | None = Field(default=None, ge=0, le=19)
    freq_num: int | None = Field(default=None, ge=1, le=366)
    freq_den: int | None = Field(default=None, ge=1, le=366)
    reminder_hour: int | None = Field(default=None, ge=0, le=23)
    reminder_min: int | None = Field(default=None, ge=0, le=59)
    reminder_days: int | None = Field(default=None, ge=0, le=127)
    target_type: int | None = Field(default=None, ge=0, le=1)
    target_value: float | None = Field(default=None, ge=0)
    unit: str | None = Field(default=None, max_length=50)
    clear_reminder: bool = False


class HabitOut(HabitBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: str
    position: int
    created_at: datetime


class HabitOverviewOut(BaseModel):
    habit: HabitOut
    score: float
    streak: int
    entries: dict[date, int]
    notes: dict[date, str]


class ReorderRequest(BaseModel):
    ordered_ids: list[int]


class EntryUpsert(BaseModel):
    value: int
    notes: str | None = Field(default=None, max_length=2000)


class EntryChangeOut(BaseModel):
    value: int
    score: float
    streak: int
    entries: dict[date, int]


class EntryOut(BaseModel):
    date: date
    value: int
    notes: str
